File: src/main/java/katrix/journeyToGensokyo/JourneyToGensokyo.java
/**
 * This class was created by <Katrix>. It's distributed as
 * part of the Journey To Gensokyo Mod. Get the Source Code in github:
 * https://github.com/Katrix-/JTG
 *
 * Journey To Gensokyo is Open Source and distributed under the
 * a modifed Botania license: https://github.com/Katrix-/JTG/blob/master/LICENSE.md
 */

package katrix.journeyToGensokyo;

import cpw.mods.fml.common.Loader;
import cpw.mods.fml.common.Mod;
import cpw.mods.fml.common.Mod.EventHandler;
import cpw.mods.fml.common.Mod.Instance;
import cpw.mods.fml.common.SidedProxy;
import cpw.mods.fml.common.event.FMLInitializationEvent;
import cpw.mods.fml.common.event.FMLPostInitializationEvent;
import cpw.mods.fml.common.event.FMLPreInitializationEvent;
import katrix.journeyToGensokyo.block.JTGBlock;
import katrix.journeyToGensokyo.handler.ChestGenHandler;
import katrix.journeyToGensokyo.handler.ConfigHandler;
import katrix.journeyToGensokyo.handler.OreDictionaryHandler;
import katrix.journeyToGensokyo.item.JTGItem;
import katrix.journeyToGensokyo.item.crafting.CraftingIC2;
import katrix.journeyToGensokyo.item.crafting.CraftingVanilla;
import katrix.journeyToGensokyo.lib.LibMod;
import katrix.journeyToGensokyo.net.PacketHandler;
import katrix.journeyToGensokyo.plugin.botania.JTGBotania;
import katrix.journeyToGensokyo.plugin.thaumcraft.JTGThaumcraft;
import katrix.journeyToGensokyo.plugin.thsc.JTG_THSC;
import katrix.journeyToGensokyo.util.LogHelper;
import katrix.journeyToGensokyo.worldgen.JTGWorldGen;

@Mod(modid = LibMod.MODID, name = LibMod.NAME, version = LibMod.VERSION, guiFactory = LibMod.GUI_FACTORY_CLASS, dependencies = "required-after:THKaguyaMod;"
		+ "after:Thaumcraft;" + "after:IC2;" + "after:Botania;" + "after:ThermalExpansion;" + "after:AdvancedSolarPanel")

public class JourneyToGensokyo {

	public static boolean IC2Installed; 
	public static boolean thaumcraftInstalled;
	public static boolean botaniaInstalled;

	@Instance
	public static JourneyToGensokyo instance;

	@SidedProxy(clientSide = LibMod.CLIENT_PROXY_CLASS, serverSide = LibMod.SERVER_PROXY_CLASS)
	public static CommonProxy proxy;

	@EventHandler
	public void preInit(FMLPreInitializationEvent event) {

		ConfigHandler.setConfig(event.getSuggestedConfigurationFile());
		
		IC2Installed = Loader.isModLoaded("IC2");
		thaumcraftInstalled = Loader.isModLoaded("Thaumcraft");
		botaniaInstalled = Loader.isModLoaded("Botania");

		PacketHandler.preInit();

		JTGItem.preInit();
		JTGBlock.preInit();
		JTG_THSC.preInit();
	}

	@EventHandler
	public void load(FMLInitializationEvent event) {

		OreDictionaryHandler.init();

		proxy.registerRenderers();

		if (IC2Installed && ConfigHandler.OresEnabled) {
			LogHelper.info("JTG adding IC2 Ore recipes");
			CraftingIC2.init();
		}
		
		CraftingVanilla.misc();
		CraftingVanilla.smelting();
		CraftingVanilla.notes();

		if (ConfigHandler.NotesEnabled) {
			LogHelper.warn("JTG Notes enabled, this is not a officialy suported feature");
			ChestGenHandler.init();
		}

		if (botaniaInstalled) {
			JTGBotania.init();
		}

		JTGWorldGen.init();

	}

	@EventHandler
	public void postInit(FMLPostInitializationEvent event) {

		JTG_THSC.postInit();
		if (thaumcraftInstalled) {
			LogHelper.info("JTG adding Thaumcraft aspects");
			JTGThaumcraft.postInit();
		}
	}
}

File: src/main/scala/net/katsstuff/journeytogensokyo/JourneyToGensokyo.scala
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the Journey To Gensokyo Mod. Get the Source Code in github:
 * https://github.com/Katrix-/JTG
 *
 * Journey To Gensokyo is Open Source and distributed under the
 * a modifed Botania license: https://github.com/Katrix-/JTG/blob/devDanmakuCore/LICENSE.md
 */
package net.katsstuff.journeytogensokyo

import net.katsstuff.journeytogensokyo.client.ClientProxy
import net.katsstuff.journeytogensokyo.handler.{CameraHandler, GuiHandler, LootHandler}
import net.katsstuff.journeytogensokyo.helper.LogHelper
import net.katsstuff.journeytogensokyo.lib.{LibMod, LibModJ}
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.fml.common.Mod.EventHandler
import net.minecraftforge.fml.common.event.{FMLInitializationEvent, FMLPreInitializationEvent}
import net.minecraftforge.fml.common.network.NetworkRegistry
import net.minecraftforge.fml.common.{FMLCommonHandler, Mod, SidedProxy}
import net.minecraftforge.fml.relauncher.Side

@Mod(
  modid = LibMod.Id,
  name = LibMod.Name,
  version = LibMod.Version,
  modLanguage = "scala",
  dependencies = "required-after:danmakucore"
)
object JourneyToGensokyo {
  MinecraftForge.EVENT_BUS.register(CommonProxy)

  if (FMLCommonHandler.instance().getSide == Side.CLIENT) {
    MinecraftForge.EVENT_BUS.register(ClientProxy)
    MinecraftForge.EVENT_BUS.register(CameraHandler)
  }

  assert(LibMod.Id == LibModJ.ID)

  //noinspection VarCouldBeVal
  @SidedProxy(serverSide = LibMod.CommonProxy, clientSide = LibMod.ClientProxy, modId = LibMod.Id)
  var proxy: CommonProxy = _

  @EventHandler
  def preInit(event: FMLPreInitializationEvent): Unit = {
    LogHelper.assignLog(event.getModLog)
    val lootHandler = new LootHandler
    lootHandler.registerLootTables()
    MinecraftForge.EVENT_BUS.register(lootHandler)

    proxy.registerRenderers()
  }

  @EventHandler
  def init(event: FMLInitializationEvent): Unit = {
    NetworkRegistry.INSTANCE.registerGuiHandler(this, GuiHandler)

    proxy.registerDanmakuCrafting()
    proxy.registerEntities()
    proxy.registerCrafting()
    proxy.registerWorldGen()
  }
}
